javascript/src/ynotebook.ts

```diff
@@ -0,0 +1,567 @@
+ *
+ * Shared cells can be inserted into a SharedNotebook.
+ * Shared cells only start emitting events when they are connected to a SharedNotebook.
+ *
+ * "Standalone" cells must not be inserted into a (Shared)Notebook.
+ * Standalone cells emit events immediately after they have been created, but they must not
+ * be included into a (Shared)Notebook.
+ */
+export class YNotebook
+  extends YDocument<NotebookChange>
+  implements ISharedNotebook
+{
+  /**
+   * Create a new notebook
+   *
+   * #### Notes
+   * The document is empty and must be populated
+   *
+   * @param options
+   */
+  constructor(options: ISharedNotebook.IOptions = {}) {
+    super();
+    this._disableDocumentWideUndoRedo =
+      options.disableDocumentWideUndoRedo ?? false;
+    this.cells = this._ycells.toArray().map(ycell => {
+      if (!this._ycellMapping.has(ycell)) {
+        this._ycellMapping.set(
+          ycell,
+          createCellModelFromSharedType(ycell, { notebook: this })
+        );
+      }
+      return this._ycellMapping.get(ycell) as YCellType;
+    });
+
+    this.undoManager.addToScope(this._ycells);
+    this._ycells.observe(this._onYCellsChanged);
+    this.ymeta.observeDeep(this._onMetaChanged);
+  }
+
+  /**
+   * YJS map for the notebook metadata
+   */
+  readonly ymeta: Y.Map<any> = this.ydoc.getMap('meta');
+  /**
+   * Cells list
+   */
+  readonly cells: YCellType[];
+
+  /**
+   * Wether the undo/redo logic should be
+   * considered on the full document across all cells.
+   *
+   * Default: false
+   */
+  get disableDocumentWideUndoRedo(): boolean {
+    return this._disableDocumentWideUndoRedo;
+  }
+
+  /**
+   * Notebook metadata
+   */
+  get metadata(): nbformat.INotebookMetadata {
+    return this.getMetadata();
+  }
+  set metadata(v: nbformat.INotebookMetadata) {
+    this.setMetadata(v);
+  }
+
+  /**
+   * Signal triggered when a metadata changes.
+   */
+  get metadataChanged(): ISignal<this, IMapChange> {
+    return this._metadataChanged;
+  }
+
+  /**
+   * nbformat major version
+   */
+  get nbformat(): number {
+    return this.ymeta.get('nbformat');
+  }
+  set nbformat(value: number) {
+    this.transact(() => {
+      this.ymeta.set('nbformat', value);
+    }, false);
+  }
+
+  /**
+   * nbformat minor version
+   */
+  get nbformat_minor(): number {
+    return this.ymeta.get('nbformat_minor');
+  }
+  set nbformat_minor(value: number) {
+    this.transact(() => {
+      this.ymeta.set('nbformat_minor', value);
+    }, false);
+  }
+
+  /**
+   * Dispose of the resources.
+   */
+  dispose(): void {
+    if (this.isDisposed) {
+      return;
+    }
+    this._ycells.unobserve(this._onYCellsChanged);
+    this.ymeta.unobserveDeep(this._onMetaChanged);
+    super.dispose();
+  }
+
+  /**
+   * Creates the YNotebook schema
+   *
+   * Note: This method is useful when we need to initialize
+   * the YNotebook from the JavaScript side.
+   */
+  initialize(): void {
```

Comment:
I would go for a generalizing a `static createStandalone` method for all `YDocument` and not introduce an `initialize` method that raises to many questions about when to call it.